decisions/C-ManagerInterface.md

```diff
@@ -0,0 +1,205 @@
+we can bundle a string buffer and length in a single C struct for error
+messages and parameters, e.g.
+
+```c++
+typedef struct {
+  const size_t maxSize;
+  char* buffer;
+  size_t usedSize;
+} OPENASSETIO_NS(SimpleString);
+```
+`usedSize` is for optimisation and safety and can technically be removed
+if we're happy to assume null-termination and accept the performance
+implications.
+
+## Opaque handles
+
+The opaque handle pattern used commonly in C APIs involves an arbitrary
+pointer to some user data, often coded simply as a `void*`. To add some
+meagre type safety, we can define an opaque handle as a pointer to an
+incomplete and ultimately unused class. We then `reinterpret_cast` it 
+to/from our actual pointer type on the C++ side. E.g.
+
+```c++
+typedef struct OPENASSETIO_NS(managerAPI_ManagerInterface_t) *
+    OPENASSETIO_NS(managerAPI_ManagerInterface_h);
+```
+
+The type `managerAPI_ManagerInterface_t` is never used in practice, its
+name is unimportant (but obviously cannot conflict with other symbols).
+It is the `managerAPI_ManagerInterface_h` "handle" (pointer) type that
+we will `reinterpret_cast` to/from.
+
+## Function pointer suites
+
+Once we have a handle to our instance, we need functions that take a 
+handle as a parameter, operate on it, then potentially return a 
+value or error. 
+
+These can be defined as top-level functions exported as independent
+symbols, appropriately namespaced.  
+
+An alternative is to bundle the functions related to a particular handle
+type into a struct of function pointers, which we call a function 
+pointer _suite_.
+
+Such suites have a few advantages
+* Aesthetically, a struct of function pointers mirrors a class with
+  methods, and if the suite is providing a C API to a C++ class then
+  mentally mapping one to the other is straightforward.
+* Only the suite type needs namespacing, the "methods" are simply
+  struct members, which reduces code verbosity.
+* Since a suite is an instance of a struct, it must be provided by
+  a function call, allowing the provider a great deal of flexibility
+  in its construction, compared to link-time defined symbols.
+* The suite is an object that can be passed around, including as input
+  to a wrapper C++ class (see below).
+* In C++ the suite's functions can be implemented as a series of 
+  (non-capturing) lambdas, keeping related code grouped together.
+
+### Return values and errors
+
+C++ functions may or may not return a value and may or may not throw an
+exception. C doesn't have exceptions, so to distinguish different error
+types we must use another mechanism. Commonly (and used here) this is
+via (integer) error codes. 
+
+Given this, a C API function wrapping a C++ class method has between 0
+and 3 possible output values, depending on whether a return value is
+expected and whether an error code and message (i.e. exception) is
+possible.
```

Comment:
I wonder if it makes sense to say "(i.e. an exceptional condition)" here, to avoid the C++ term "exception". The reason is I had to read this section a couple of times to convince myself that it doesn't allow a C++ routine to throw across the C interface.

Reply:
I've re-worded slightly in 46a0a3c0 - hopefully that removes any initial confusion.